posts/2021-06-11-request-timing-21006.adoc

```diff
@@ -150,10 +150,7 @@ The problem was fixed by adding steps to transfer the `useJandex` setting to the
 
 * link:https://github.com/OpenLiberty/open-liberty/issues/17203[ORB.init() called simultaneously on two threads during server start]
 +
-We found that Java’s `AtomicReference.getAndUpdate()` did not prevent the initialization code from running twice.
-(This is the design of AtomicReference but a false assumption about this led to a bug in Liberty.)
-To avoid this, we wrote our own concurrent utility, AcidReference, that guaranteed a single thread would win the race to perform the initialization.
-Consequently, when two components are starting up in parallel in Liberty and both call for the `ORB`, only one of the threads performs the initialization and the other one waits for the result.
+We found a problem in Liberty's ORB initialization code, based on a false assumuption that Java's `AtomicReference.getAndUpdate()` would only call the provided 'updateFunction' once for any update to the reference's value. (In fact, though it guarantees the update to the reference's value to be atomic, the `updateFunction` might be called by several threads contending to perform the update). To provide the behaviour we required, we devised our own concurrent utility, AcidReference, (built using AtomicReference), which guarantees the `updateFunction` is only called once, by one of the threads contending for update. Making use of this new `AcidReference` allows any contention in ORB initialization in Liberty to be resolved correctly, tasking one thread to perform the ORB initialization whilst any others wait for its result.
```

Comment:
Suggested edit:

We found a problem in the Liberty ORB initialization code, based on a false assumuption that the Java `AtomicReference.getAndUpdate()` would only call the provided 'updateFunction' once for any update to the reference value. In fact, though it guarantees the update to the reference's value to be atomic, the `updateFunction` might be called by several threads contending to perform the update. To provide the behavior we required, we devised our own concurrent utility, AcidReference, which was built by using AtomicReference and  guarantees the `updateFunction` is only called once, by one of the threads contending for update. Making use of this new `AcidReference` allows any contention in ORB initialization in Liberty to be resolved correctly, tasking one thread to perform the ORB initialization while any others wait for its result.